pennylane/tape/qscript.py

```diff
@@ -0,0 +1,1207 @@
+
+    Args:
+        ops (Iterable[Operator]): An iterable of the operations to be performed
+        measurements (Iterable[MeasurementProcess]): All the measurements to be performed
+        prep (Iterable[Operator]): Any state preparations to perform at the start of the circuit
+        name (str): a name given to the quantum tape
+
+    .. seealso:: :class:`pennylane.tape.QuantumTape`
+
+    **Example:**
+
+    .. code-block:: python
+
+        from pennylane.qscript import QuantumScript
+
+        prep = [qml.BasisState([1,1], wires=(0,1))]
+
+        ops = [qml.RX(0.432, 0),
+               qml.RY(0.543, 0),
+               qml.CNOT((0,"a")),
+               qml.RX(0.133, "a")]
+
+        qscript = QuantumScript(ops, [qml.expval(qml.PauliZ(0))], prep)
+
+    >>> list(qscript)
+    [RX(0.432, wires=[0]),
+    RY(0.543, wires=[0]),
+    CNOT(wires=[0, 'a']),
+    RX(0.133, wires=['a']),
+    expval(PauliZ(wires=[0]))]
+    >>> qscript.operations
+    [RX(0.432, wires=[0]),
+    RY(0.543, wires=[0]),
+    CNOT(wires=[0, 'a']),
+    RX(0.133, wires=['a'])]
+    >>> qscript.measurements
+    [expval(PauliZ(wires=[0]))]
+
+    Iterating over the quantum script can be done by:
+
+    >>> for op in qscript:
+    ...     print(op)
+    RX(0.432, wires=[0])
+    RY(0.543, wires=[0])
+    CNOT(wires=[0, 'a'])
+    RX(0.133, wires=['a'])
+    expval(PauliZ(wires=[0]))'
+
+    Quantum scripts also support indexing and length determination:
+
+    >>> qscript[0]
+    RX(0.432, wires=[0])
+    >>> len(qscript)
+    5
+
+    Once constructed, the script can be executed directly on a quantum device
+    using the :func:`~.pennylane.execute` function:
+
+    >>> dev = qml.device('default.qubit', wires=(0,'a'))
+    >>> qml.execute([qscript], dev, gradient_fn=None)
+    [array([0.77750694])]
+
+    ``ops``, ``measurements``, and ``prep`` are converted to lists upon initialization,
+    so those arguments accept any iterable object:
+
+    >>> qscript = QuantumScript((qml.PauliX(i) for i in range(3)))
+    >>> qscript.circuit
+    [PauliX(wires=[0]), PauliX(wires=[1]), PauliX(wires=[2])]
+
+    """
+
+    def __init__(self, ops=None, measurements=None, prep=None, name=None, _update=True):
+        # are we going to queue these?
+        self.name = name
+        self._prep = [] if prep is None else list(prep)
+        self._ops = [] if ops is None else list(ops)
+        self._measurements = [] if measurements is None else list(measurements)
+
+        self._par_info = {}
+        """dict[int, dict[str, Operator or int]]: Parameter information. Keys are
+        parameter indices (in the order they appear on the tape), and values are a
+        dictionary containing the corresponding operation and operation parameter index."""
+
+        self._trainable_params = []
+        self._graph = None
+        self._specs = None
+        self._output_dim = 0
+        self._batch_size = None
+        self._qfunc_output = None
+
+        self.wires = _empty_wires
```

Comment:
Same thing.  I think this change was made so that we don't have to keep initializing empty wires.  

We should actually avoid this by directly using the code from `_update_circuit_info`.  I wonder how much I should improve the code in this PR, or just copy over all the code with minimal changes, and improve it in a later PR?

Reply:
I prefer separating it into different PRs